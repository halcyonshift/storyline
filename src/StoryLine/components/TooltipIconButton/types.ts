import { MouseEvent, ReactElement } from 'react'
import { IconButtonProps } from '@mui/material'

export type TooltipIconButtonProps = {
    text: string
    icon: ReactElement
    link?: string
    confirm?: string
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    onClick?: (e?: MouseEvent<HTMLButtonElement, MouseEvent>) => Promise<void> | void
} & IconButtonProps
