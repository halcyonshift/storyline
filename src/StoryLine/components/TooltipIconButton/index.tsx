import { useState } from 'react'
import {
    Button,
    Dialog,
    DialogActions,
    DialogContent,
    DialogContentText,
    IconButton,
    Tooltip
} from '@mui/material'
import { useTranslation } from 'react-i18next'
import { useNavigate } from 'react-router-dom'

import { TooltipIconButtonProps } from './types'

const TooltipIconButton = ({
    text,
    icon,
    link,
    onClick,
    confirm,
    ...props
}: TooltipIconButtonProps) => {
    const { t } = useTranslation()
    const navigate = useNavigate()

    const [open, setOpen] = useState<boolean>(false)

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const handleAction = async (e: any) => {
        if (onClick) await onClick(e)
        if (link) navigate(link)
        setOpen(false)
    }

    return (
        <>
            <Tooltip title={t(text)}>
                <IconButton
                    color='inherit'
                    onClick={async (e) => {
                        if (confirm) {
                            return setOpen(true)
                        }
                        handleAction(e)
                    }}
                    aria-label={t(text)}
                    {...props}>
                    {icon}
                </IconButton>
            </Tooltip>
            <Dialog
                open={open}
                onClose={() => setOpen(false)}
                aria-describedby='alert-dialog-description'>
                <DialogContent>
                    <DialogContentText id='alert-dialog-description'>{confirm}</DialogContentText>
                </DialogContent>
                <DialogActions>
                    <Button onClick={() => setOpen(false)}>
                        {t('component.tooltipIconButton.dialog.no')}
                    </Button>
                    <Button color='error' onClick={handleAction} autoFocus>
                        {t('component.tooltipIconButton.dialog.yes')}
                    </Button>
                </DialogActions>
            </Dialog>
        </>
    )
}

export default TooltipIconButton
