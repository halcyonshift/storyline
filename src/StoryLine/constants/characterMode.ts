export type CharacterModeType = 'PRIMARY' | 'SECONDARY' | 'TERTIARY'

export const CharacterMode = Object.freeze({
    PRIMARY: 'PRIMARY',
    SECONDARY: 'SECONDARY',
    TERTIARY: 'TERTIARY'
})
