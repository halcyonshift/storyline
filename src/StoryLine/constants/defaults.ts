export const DEFAULT_DISPLAY_MODE = 'light'
export const DEFAULT_FONT = 'roboto'
export const DEFAULT_FONT_SIZE = 14
export const DEFAULT_INDENT_PARAGRAPH = true
export const DEFAULT_LANGUAGE = 'en'
export const DEFAULT_LINE_SPACING = 1.5
export const DEFAULT_PARAGRAPH_SPACING = 2
export const DEFAULT_SPELL_CHECK = true
export const DEFAULT_AUTO_SAVE = true
export const DEFAULT_AUTO_BACKUP_FREQ = 0
export const DEFAULT_AUTO_BACKUP_MAX = 10
export const DEFAULT_PALETTE = 'indigo'
