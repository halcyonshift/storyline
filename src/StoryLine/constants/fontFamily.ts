export const FontFamily = Object.freeze({
    arial: 'Arial',
    Avenir: 'Avenir',
    baskerville: 'Baskerville',
    'courier new': 'Courier New',
    OpenDyslexic: 'Open Dyslexic',
    roboto: 'Roboto',
    'times new roman': 'Times New Roman',
    verdana: 'Verdana'
})
