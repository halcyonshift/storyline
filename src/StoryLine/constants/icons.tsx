import Icon from '@sl/components/Icon'

export const GLOBAL_ICONS = {
    add: <Icon name='Add' />,
    addImage: <Icon name='AddPhotoAlternate' />,
    back: <Icon name='ArrowBackIosNew' />,
    change: <Icon name='ChangeCircleOutlined' />,
    close: <Icon name='Close' />,
    delete: <Icon name='Delete' />,
    down: <Icon name='ArrowDownward' />,
    edit: <Icon name='EditOutlined' />,
    expand: <Icon name='ExpandMore' />,
    history: <Icon name='History' />,
    info: <Icon name='Info' />,
    next: <Icon name='ArrowForwardIos' />,
    open: <Icon name='FileOpen' />,
    random: <Icon name='Casino' />,
    up: <Icon name='ArrowUpward' />
}

export const CHARACTER_ICONS = {
    character: <Icon name='Group' />,
    addPrimary: <Icon name='PersonAddAlt1' />,
    addSecondary: <Icon name='PersonAddAlt' />,
    addTertiary: <Icon name='GroupAdd' />,
    primary: <Icon name='Person' />,
    secondary: <Icon name='PersonOutline' />,
    tertiary: <Icon name='PeopleAlt' />
}

export const CONNECTION_ICONS = {
    connections: <Icon name='Hub' />
}

export const IMPORTEXPORT_ICONS = {
    docx: <Icon name='Description' />,
    html: <Icon name='Html' />,
    importExport: <Icon name='ImportExport' />,
    import: <Icon name='ArrowDownward' />,
    pdf: <Icon name='PictureAsPdf' />
}

export const INSIGHT_ICONS = {
    insights: <Icon name='Insights' />
}

export const ITEM_ICONS = {
    item: <Icon name='Category' />,
    add: <Icon name='AddCircle' />
}

export const LOCATION_ICONS = {
    location: <Icon name='LocationOn' />,
    add: <Icon name='AddLocationAlt' />
}

export const NOTE_ICONS = {
    note: <Icon name='StickyNote2' />,
    add: <Icon name='PostAdd' />
}

export const OVERVIEW_ICONS = {
    summary: <Icon name='FormatListBulleted' />,
    filter: <Icon name='FilterList' />
}

export const RICHTEXT_ICONS = {
    bold: <Icon name='FormatBold' />,
    italic: <Icon name='FormatItalic' />,
    ul: <Icon name='FormatListBulleted' />,
    ol: <Icon name='FormatListNumbered' />,
    strike: <Icon name='FormatStrikethrough' />,
    underline: <Icon name='FormatUnderlined' />,
    alignLeft: <Icon name='FormatAlignLeft' />,
    alignRight: <Icon name='FormatAlignRight' />,
    alignCenter: <Icon name='FormatAlignCenter' />,
    alignJustify: <Icon name='FormatAlignJustify' />,
    tag: <Icon name='LocalOffer' />,
    version: <Icon name='RestorePage' />,
    redo: <Icon name='Redo' />,
    undo: <Icon name='Undo' />,
    save: <Icon name='Save' />,
    search: <Icon name='Search' />,
    excerpt: <Icon name='LabelImportant' />
}

export const SEARCH_ICONS = {
    search: <Icon name='Search' />,
    clear: <Icon name='Clear' />
}

export const SECTION_ICONS = {
    section: <Icon name='MenuBook' />,
    addPart: <Icon name='Queue' />,
    addChapter: <Icon name='LibraryBooks' />,
    addScene: <Icon name='PeopleAlt' />,
    version: <Icon name='RestorePage' />
}

export const SETTINGS_ICONS = {
    settings: <Icon name='Settings' />
}

export default {
    global: GLOBAL_ICONS,
    character: CHARACTER_ICONS,
    connection: CONNECTION_ICONS,
    importExport: IMPORTEXPORT_ICONS,
    insight: INSIGHT_ICONS,
    item: ITEM_ICONS,
    location: LOCATION_ICONS,
    note: NOTE_ICONS,
    overview: OVERVIEW_ICONS,
    search: SEARCH_ICONS,
    section: SECTION_ICONS,
    settings: SETTINGS_ICONS
}
