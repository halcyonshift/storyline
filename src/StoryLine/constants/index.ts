export const YYYYMMDD = 'yyyy-MM-dd'
