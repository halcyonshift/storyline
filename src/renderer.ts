import { init } from '@sentry/electron/renderer'
import { init as reactInit } from '@sentry/react'

if (parseInt(process.env.DEBUG) === 0) {
    init({ dsn: process.env.SENTRY }, reactInit)
}

import '@fontsource/roboto/300.css'
import '@fontsource/roboto/400.css'
import '@fontsource/roboto/500.css'
import '@fontsource/roboto/700.css'
import './index.css'
import './StoryLine'
