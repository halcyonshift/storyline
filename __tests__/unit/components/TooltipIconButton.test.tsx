import userEvent from '@testing-library/user-event'
import renderer from 'react-test-renderer'
import TooltipIconButton from '@sl/components/TooltipIconButton'
import { GLOBAL_ICONS } from '@sl/constants/icons'
import { render, screen } from '../test-utils'

const onClick = jest.fn()
const mockedUsedNavigate = jest.fn()

jest.mock('react-router-dom', () => ({
    ...(jest.requireActual('react-router-dom') as any),
    useNavigate: () => mockedUsedNavigate
}))

describe('<TooltipIconButton />', () => {
    it('has the correct text and icon', async () => {
        render(<TooltipIconButton text='The text' icon={GLOBAL_ICONS.add} />)
        expect(screen.getByLabelText('The text')).toBeTruthy()
        expect(screen.getByLabelText('Add')).toBeTruthy()
    })

    it('navigates to the correct link', async () => {
        render(<TooltipIconButton text='The text' icon={GLOBAL_ICONS.add} link='/a-link' />)
        await userEvent.click(screen.getByRole('button'))
        expect(mockedUsedNavigate).toHaveBeenCalledWith('/a-link')
    })

    it('handles the onClick with no confirm', async () => {
        render(<TooltipIconButton text='The text' icon={GLOBAL_ICONS.add} onClick={onClick} />)
        await userEvent.click(screen.getByRole('button'))
        expect(onClick).toHaveBeenCalled()
    })

    it('handles the onClick with confirm', async () => {
        render(
            <TooltipIconButton
                text='The text'
                icon={GLOBAL_ICONS.add}
                onClick={onClick}
                link='/a-link'
                confirm='CONFIRM'
            />
        )
        const closedButtons = await screen.findAllByRole('button')
        expect(closedButtons.length).toBe(1)

        await userEvent.click(screen.getByRole('button'))
        const dialog = await screen.findByText('CONFIRM')

        expect(dialog).toBeTruthy()

        const openButtons = await screen.findAllByRole('button')
        await userEvent.click(openButtons[2])

        expect(onClick).toHaveBeenCalled()
        expect(mockedUsedNavigate).toHaveBeenCalledWith('/a-link')
    })

    it('renders correctly', () => {
        const tree = renderer
            .create(
                <TooltipIconButton
                    text='The text'
                    icon={GLOBAL_ICONS.add}
                    onClick={onClick}
                    link='/a-link'
                    confirm='CONFIRM'
                />
            )
            .toJSON()
        expect(tree).toMatchSnapshot()
    })
})
