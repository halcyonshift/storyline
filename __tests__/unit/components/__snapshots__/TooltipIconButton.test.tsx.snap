// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TooltipIconButton /> renders correctly 1`] = `
<button
  aria-label="The text"
  aria-labelledby={null}
  className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
  data-mui-internal-clone-element={true}
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <svg
    aria-hidden={true}
    aria-label="Add"
    className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
    data-testid="AddOutlinedIcon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
    />
  </svg>
  <span
    className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;
